crates/oxc_linter/src/rules/promise/no_callback_in_promise.rs

```diff
@@ -0,0 +1,193 @@
+use oxc_ast::{
+    ast::{Expression, MemberExpression},
+    AstKind,
+};
+use oxc_diagnostics::OxcDiagnostic;
+use oxc_macros::declare_oxc_lint;
+use oxc_span::{GetSpan, Span};
+
+use crate::{context::LintContext, rule::Rule, AstNode};
+
+static CALLBACKS: [&str; 4] = ["done", "cb", "callback", "next"];
+
+fn no_callback_in_promise_diagnostic(span: Span) -> OxcDiagnostic {
+    OxcDiagnostic::warn("Avoid calling back inside of a promise").with_label(span)
+}
+
+#[derive(Debug, Default, Clone)]
+pub struct NoCallbackInPromise(Box<NoCallbackInPromiseConfig>);
+
+#[derive(Debug, Default, Clone)]
+pub struct NoCallbackInPromiseConfig {
+    exceptions: Vec<String>,
+}
+
+impl std::ops::Deref for NoCallbackInPromise {
+    type Target = NoCallbackInPromiseConfig;
+
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+declare_oxc_lint!(
+    /// ### What it does
+    ///
+    /// Disallows calling a callback function (`cb()`) inside a `Promise.prototype.then()`
+    /// or `Promise.prototype.catch()`.
+    ///
+    /// ### Why is this bad?
+    ///
+    /// Directly invoking a callback inside a `then()` or `catch()` method can lead to
+    /// unexpected behavior, such as the callback being called multiple times. Additionally,
+    /// mixing the callback and promise paradigms in this way can make the code confusing
+    /// and harder to maintain.
+    ///
+    /// ### Examples
+    ///
+    /// Examples of **incorrect** code for this rule:
+    /// ```js
+    /// function callback(err, data) {
+    ///     console.log('Callback got called with:', err, data)
+    ///     throw new Error('My error')
+    ///   }
+    ///
+    /// Promise.resolve()
+    ///   .then(() => callback(null, 'data'))
+    ///   .catch((err) => callback(err.message, null))
+    /// ```
+    ///
+    /// Examples of **correct** code for this rule:
+    /// ```js
+    /// Promise.resolve()
+    ///   .then(() => setTimeout(() => callback(null, 'data'), 0))
+    ///   .catch((err) => setTimeout(() => callback(err.message, null), 0))
+    /// ```
+    NoCallbackInPromise,
+    correctness,
+);
+
+impl Rule for NoCallbackInPromise {
+    fn from_configuration(value: serde_json::Value) -> Self {
+        Self(Box::new(NoCallbackInPromiseConfig {
+            exceptions: value
+                .get(0)
+                .and_then(|v| v.get("exceptions"))
+                .and_then(serde_json::Value::as_array)
+                .map(|v| {
+                    v.iter()
+                        .filter_map(serde_json::Value::as_str)
+                        .map(ToString::to_string)
```

Comment:
Do we need to allocate on the heap here? Could we use `CompactStr` to store these exceptions instead? For example: https://github.com/oxc-project/oxc/blob/40bd9198496dbf2dbcbff1b1cb58869aa3443e92/crates/oxc_linter/src/rules/eslint/no_extend_native.rs#L76-L83

Reply:
Let me try to attempt it